tensorpack/utils/nvml.py

```diff
@@ -0,0 +1,290 @@
+                  "17": "NVML_ERROR_OPERATING_SYSTEM",
+                  "18": "NVML_ERROR_LIB_RM_VERSION_MISMATCH",
+                  "999": "NVML_ERROR_UNKNOWN"}
+
+
+def NvmlCodeToString(i):
+    return NvmlErrorCodes[str(i)]
+
+
+class NvmlException(Exception):
+    def __init__(self, error_code):
+        super(NvmlException, self).__init__(error_code)
+        self.error_code = error_code
+
+    def __str__(self):
+        return NvmlCodeToString(self.error_code)
+
+
+def CheckNvmlReturn(ret):
+    if (ret != 0):
+        raise NvmlException(ret)
+    return ret
+
+
+class NVML(object):
+
+    def load(self):
+        global nvmlLib
+        _lib_lock.acquire()
+        try:
+            nvmlLib = CDLL("libnvidia-ml.so.1")
+        except OSError:
+            NvmlException(NVML_ERROR_LIBRARY_NOT_FOUND)
+        finally:
+            _lib_lock.release()
+
+        self.build_cache()
+
+    def function_pointer(self, name):
+        _lib_lock.acquire()
+        try:
+            try:
+                return getattr(nvmlLib, name)
+            except AttributeError:
+                raise NvmlException(NVML_ERROR_FUNCTION_NOT_FOUND)
+        finally:
+            _lib_lock.release()
+
+    def build_cache(self):
+        function_pointers = ["nvmlDeviceGetName", "nvmlDeviceGetUUID", "nvmlDeviceGetMemoryInfo",
+                             "nvmlDeviceGetUtilizationRates", "nvmlInit_v2", "nvmlShutdown",
+                             "nvmlDeviceGetCount_v2", "nvmlDeviceGetHandleByIndex_v2"]
+
+        self.func_ptr = {n: self.function_pointer(
+            n) for n in function_pointers}
+
+    def get_function(self, name):
+        if name in self.func_ptr.keys():
+            return self.func_ptr[name]
+
+    def call(self, hnd, buf_len, func_ptr):
+        c_name = create_string_buffer(buf_len)
+        # fn = function_pointer_cache[func_ptr]
+        fn = self.func_ptr[func_ptr]
+        ret = fn(hnd, c_name, c_uint(buf_len))
+        CheckNvmlReturn(ret)
+        return c_name
+
+
+_NVML = NVML()
+
+
+class GpuDevice(Structure):
+    """Represent GPU Information
+    """
+    pass
+
+
+c_nvmlDevice_t = POINTER(GpuDevice)
+
+
+class NvidiaDevice(object):
+    """Represent a single GPUDevice"""
+
+    def __init__(self, hnd):
+        super(NvidiaDevice, self).__init__()
+        self.hnd = hnd
+
+    def Name(self):
+        """Return GPU name
+
+        Example:
+
+            >>> print(nvidia.Device(0).Name())
+            GeForce GTX 970
+
+        Returns:
+            Name of GPU brand
+        """
+        return _NVML.call(self.hnd, NVML_DEVICE_NAME_BUFFER_SIZE, "nvmlDeviceGetName").value
+
+    def UUID(self):
+        """Return GPU UUID
+
+        Example:
+
+            >>> print(nvidia.Device(0).UUID())
+            GPU-xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+
+        Returns:
+            UUID of GPU
+        """
+        return _NVML.call(self.hnd, NVML_DEVICE_UUID_BUFFER_SIZE, "nvmlDeviceGetUUID").value
```

Comment:
I don't think UUID and name of card will ever be useful. They can be removed together with call() and BUFFER constants.